### DISCUSSION TITLE
How to get the request body and service response body in a kong custom plugin?

### DISCUSSION THREAD

**Discussion by shahamit** [Kong/kubernetes-ingress-controller#1876] [Q&A] [answered]
I am trying to implement a custom kong plugin that would encrypt the (proxied) service response before it returns it to the client and decrypt the request body before it passes it to the proxied service.

To do so I am trying to fetch the request body and the response body and set it as a new header value for testing purposes. I am unable to fetch them though. I get `curl: (52) Empty reply from server` without any error in the ingress-kong proxy container.
I tried to come up with the below code after following [this github plugin](https://github.com/wshirey/kong-plugin-jwt-claims-headers/blob/master/handler.lua) and [this api doc](https://docs.konghq.com/gateway-oss/2.5.x/pdk/kong.service.response/). It seems to the api doc is for plugins created when kong is deployed as a standalone api gateway. For kong kubernetes ingress controller, do we need to replace the prefix kong to ngx? Though I didn't follow the reasoning but this also partially worked as demo'd below

Could you please guide me on how to access the request body and service.response.body in the lua handler code? 

Thanks 

My lua handler code is as follows

```
local BasePlugin = require "kong.plugins.base_plugin"

local APIEncryptDecrypt = {
  PRIORITY = 1000, 
  VERSION = "0.1",
}

local APIEncryptDecrypt = BasePlugin:extend()


function APIEncryptDecrypt:header_filter(conf)
  
 -- This works. I could see a new header with the appropriate value
  kong.response.set_header("myheader", ngx.req.get_headers()["Authorization"])       

--  Adding this leads to the curl error
  kong.response.set_header("from_request", ngx.service.res.get_raw_body())             

-- This also fails when added
  -- kong.response.set_header("myheader", kong.service.response.get_body())            
end

return APIEncryptDecrypt
```

**Comment by shaneutt** [comment, marked as answer]
Hi @shahamit,

The following links provide a rough overview of the request and the how to operate on the entire response body:

- https://docs.konghq.com/gateway-oss/2.5.x/pdk/kong.request/#kongrequestget_bodymimetype-max_args
- https://github.com/Kong/kong/blob/2.6.0/kong/plugins/response-transformer/handler.lua#L23-L42

If you want to perform a response transform that the stock response transformer can't do you can follow the "aggregate each of the body_filter iterations into a table, concatenate them, and then pass into your transform function when you get an EOF" pattern from above (and probably even reuse a fair amount of the code). You'll write your own function where response-transformer otherwise uses `transform_json_body()`.

**Reply by shahamit** [reply]
Fetching the header values and the request body worked fine with the below code but fetching the service response body fails. Is there some reason why `kong.request` works but `kong.service.response` doesn't? I tried moving the failing code in the body_filter(conf) method too but it still doesn't work. I get these logs for the proxy container

```
[error] 1097#0: *270 failed to run body_filter_by_lua*: ...share/lua/5.1/kong/plugins/apiencryptdecrypt/handler.lua:26: service body is only available with buffered proxying (see: kong.service.request.enable_buffering function)
stack traceback:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/pdk/service/response.lua:307: in function 'get_body'
	...share/lua/5.1/kong/plugins/apiencryptdecrypt/handler.lua:26: in function <...share/lua/5.1/kong/plugins/apiencryptdecrypt/handler.lua:20>
	/usr/local/share/lua/5.1/kong/init.lua:275: in function 'execute_plugins_iterator'
	/usr/local/share/lua/5.1/kong/init.lua:1236: in function 'body_filter'
	body_filter_by_lua:2: in main chunk while sending to client, client: 10.42.0.0, server: kong, request: "GET /example/v1/hotels/2 HTTP/1.1", upstream: "http://10.42.1.6:8090/example/v1/hotels/2", host: "172.22.0.2"

```


Could you please guide on what could be going wrong? Thanks

     ```
     function APIEncryptDecrypt:access(conf)
             kong.response.set_header("myheader", ngx.req.get_headers()["Authorization"])
             kong.response.set_header("from_request", kong.request.get_raw_body())
             kong.response.set_header("myheader", kong.service.response.get_body())   -- fails
     end
    ```

**Reply by shaneutt** [reply]
It seems like `enable_buffering` is needed to do what you're doing in your plugin:

https://docs.konghq.com/gateway-oss/2.5.x/pdk/kong.service.request/#kongservicerequestenable_buffering
